source/dub/git.d

```diff
@@ -0,0 +1,40 @@
+// functionality to supply packages from git submodules
+module dub.git;
+
+import dub.dependency;
+import dub.internal.vibecompat.core.file;
+import dub.package_;
+import dub.packagemanager;
+
+import std.algorithm;
+import std.exception : enforce;
+import std.range;
+import std.string;
+
+/// adds the git submodules checked out in the root path as direct packages
+public void addGitSubmodules(PackageManager packageManager, NativePath rootPath) {
+	import std.process : execute;
+
+	auto rootScmPath = rootPath ~ ".git";
+	const submoduleInfo = execute([
+		"git",
+		"-C", rootPath.toNativeString,
+		"--git-dir=" ~ (rootScmPath.relativeTo(rootPath)).toNativeString,
+		"submodule", "status"]);
+
+	enforce(submoduleInfo.status == 0,
+		format("git submodule status exited with error code %s: %s", submoduleInfo.status, submoduleInfo.output));
+
+	foreach (line; submoduleInfo.output.lines) {
+		const parts = line.split(" ").map!strip.filter!(a => !a.empty).array;
+		const subPath = rootPath ~ parts[1];
```

Comment:
Can't we do sth. less fragile?

Reply:
I'm also not enthused about it, but I'm not sure how to improve on it.

Reply:
Bash code to enumerate submodules:

```bash
	git config -lz | \
		while IFS=$'\n' read -r -d '' name value
		do
			if [[ "$name" =~ ^submodule\.(.*)\.url$ || ( "$name" =~ ^submodule\.(.*)\.active$ && "$value" == true ) ]]
			then
				printf '%s\0' "${BASH_REMATCH[1]}"
			fi
		done | \
			sort -uz | \
			mapfile -d '' -t list
```

See https://git-scm.com/docs/gitsubmodules#_active_submodules for reference.

To get the submodule git dir:

```bash
git rev-parse --git-path modules/"$m"
```

where `$m` is the submodule name as per above list.

To get the submodule work dir:

```bash
git -C "$md" rev-parse --show-toplevel
```

where `md` is the git dir above.

Reply:
For this purpose probably we want to look at unregistered submodules too (i.e. submodule entries in the tree).

In which case, the equivalent of:

```bash
git ls-tree -r --full-tree HEAD -z | grep -z '^160000 ' | cut -z -d $'\t' -f 2 | xargs -0 printf '%s\n'
```
would be more appropriate.

Then, just check for a `dub.sdl` at each entry.